// @angular-material-extensions/password-strength/esm5/lib/validator/mat-password-strength-validator.js
var MatPasswordStrengthValidator = /** @class */ (function () {
    function MatPasswordStrengthValidator() {
    }
    MatPasswordStrengthValidator.prototype.isUndefinedOrEmpty = function (control) {
        if (!control || !control.value || control.value.length === 0) {
            return undefined;
        }
    };
    MatPasswordStrengthValidator.prototype.validate = function (criteria, regex) {
        var _this = this;
        var validator = function (control) {
            _this.isUndefinedOrEmpty(control);
            if (!regex.test(control.value)) {
                var failed = {};
                failed[criteria] = {
                    actualValue: control.value,
                    requiredPattern: regex
                };
                return failed;
            }
            return undefined;
        };
        return validator;
    };
    MatPasswordStrengthValidator.prototype.confirm = function (password) {
        var _this = this;
        var validator = function (control) {
            _this.isUndefinedOrEmpty(control);
            if (control.value !== password) {
                return {
                    notConfirmed: {
                        password: password,
                        passwordConfirmation: control.value
                    }
                };
            }
            return undefined;
        };
        return validator;
    };
    return MatPasswordStrengthValidator;
}());
export { MatPasswordStrengthValidator };
