// ngx-spinner/esm2015/lib/ngx-spinner.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxSpinnerComponent } from './ngx-spinner.component';
import { SafeHtmlPipe } from './safe-html.pipe';
import * as ɵngcc0 from '@angular/core';
export class NgxSpinnerModule {
}
NgxSpinnerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxSpinnerModule });
NgxSpinnerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxSpinnerModule_Factory(t) { return new (t || NgxSpinnerModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxSpinnerModule, { declarations: function () { return [NgxSpinnerComponent, SafeHtmlPipe]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxSpinnerComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxSpinnerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [NgxSpinnerComponent, SafeHtmlPipe],
                exports: [NgxSpinnerComponent]
            }]
    }], null, null); })();
